File: tools/memory/extract_facts.py
#!/usr/bin/env python3
"""Atomic fact extraction pipeline for cortana_memory_semantic.

Commands:
  extract_facts.py extract --input <file|-> [--dry-run]
  extract_facts.py batch --since-hours 24 [--dry-run]
"""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any

WORKSPACE = Path("/Users/hd/openclaw")
PSQL_BIN = "/opt/homebrew/opt/postgresql@17/bin/psql"
DB_NAME = "cortana"
PROMPT_FILE = WORKSPACE / "tools" / "memory" / "prompts" / "fact_extraction.txt"
EMBED_SCRIPT = WORKSPACE / "tools" / "embeddings" / "embed.py"
EMBED_BIN = WORKSPACE / "tools" / "embeddings" / "embed"
SESSIONS_GLOB = os.path.expanduser("~/.openclaw/agents/main/sessions/*.jsonl")

VALID_TYPES = {"fact", "preference", "event", "system_rule"}


@dataclass
class AtomicFact:
    fact_type: str
    content: str
    tags: list[str]
    people: list[str]
    projects: list[str]
    importance: float
    confidence: float
    supersedes_id: int | None = None


def sh(cmd: list[str], timeout: int = 120) -> subprocess.CompletedProcess[str]:
    return subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)


def psql(sql: str, capture: bool = False) -> str:
    proc = sh([PSQL_BIN, DB_NAME, "-q", "-v", "ON_ERROR_STOP=1", "-At", "-c", sql], timeout=180)
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or "psql failed")
    return (proc.stdout or "").strip() if capture else ""


def q(value: Any) -> str:
    if value is None:
        return "NULL"
    return "'" + str(value).replace("'", "''") + "'"


def ensure_schema() -> None:
    sql = """
ALTER TABLE cortana_memory_semantic
  ADD COLUMN IF NOT EXISTS fact_type TEXT,
  ADD COLUMN IF NOT EXISTS superseded_by BIGINT,
  ADD COLUMN IF NOT EXISTS superseded_at TIMESTAMPTZ,
  ADD COLUMN IF NOT EXISTS supersedes_id BIGINT,
  ADD COLUMN IF NOT EXISTS extraction_source TEXT,
  ADD COLUMN IF NOT EXISTS embedding_local VECTOR(384);

DO $$
BEGIN
  IF NOT EXISTS (
    SELECT 1 FROM pg_constraint
    WHERE conname='cortana_memory_semantic_superseded_by_fkey'
  ) THEN
    ALTER TABLE cortana_memory_semantic
      ADD CONSTRAINT cortana_memory_semantic_superseded_by_fkey
      FOREIGN KEY (superseded_by) REFERENCES cortana_memory_semantic(id);
  END IF;
END $$;

DO $$
BEGIN
  IF NOT EXISTS (
    SELECT 1 FROM pg_constraint
    WHERE conname='cortana_memory_semantic_supersedes_id_fkey'
  ) THEN
    ALTER TABLE cortana_memory_semantic
      ADD CONSTRAINT cortana_memory_semantic_supersedes_id_fkey
      FOREIGN KEY (supersedes_id) REFERENCES cortana_memory_semantic(id);
  END IF;
END $$;

DO $$
BEGIN
  IF EXISTS (
    SELECT 1 FROM pg_constraint
    WHERE conname='cortana_memory_semantic_fact_type_check'
  ) THEN
    ALTER TABLE cortana_memory_semantic DROP CONSTRAINT cortana_memory_semantic_fact_type_check;
  END IF;

  ALTER TABLE cortana_memory_semantic
    ALTER COLUMN fact_type SET DEFAULT 'fact';

  UPDATE cortana_memory_semantic
  SET fact_type = 'fact'
  WHERE fact_type IS NULL;

  ALTER TABLE cortana_memory_semantic
    ADD CONSTRAINT cortana_memory_semantic_fact_type_check
    CHECK (fact_type = ANY (ARRAY['fact','preference','event','system_rule','decision','rule','relationship']));
END $$;

CREATE INDEX IF NOT EXISTS idx_memory_semantic_embedding_local_hnsw
  ON cortana_memory_semantic USING hnsw (embedding_local vector_cosine_ops)
  WHERE embedding_local IS NOT NULL;
"""
    psql(sql)


def load_prompt() -> str:
    if not PROMPT_FILE.exists():
        raise FileNotFoundError(f"Missing prompt template: {PROMPT_FILE}")
    return PROMPT_FILE.read_text(encoding="utf-8")


def parse_jsonl_transcript(path: Path) -> str:
    lines: list[str] = []
    with path.open("r", encoding="utf-8", errors="ignore") as f:
        for raw in f:
            raw = raw.strip()
            if not raw:
                continue
            try:
                obj = json.loads(raw)
            except json.JSONDecodeError:
                continue
            if obj.get("type") != "message":
                continue
            msg = obj.get("message", {})
            role = msg.get("role")
            if role not in {"user", "assistant"}:
                continue
            text_parts: list[str] = []
            for chunk in msg.get("content", []):
                if isinstance(chunk, dict) and chunk.get("type") == "text" and isinstance(chunk.get("text"), str):
                    text_parts.append(chunk["text"])
            if text_parts:
                lines.append(f"{role.upper()}: {' '.join(text_parts).strip()}")
    return "\n".join(lines)


def read_input(input_path: str) -> tuple[str, str]:
    if input_path == "-":
        return "stdin", sys.stdin.read()
    p = Path(input_path)
    if not p.exists():
        raise FileNotFoundError(input_path)
    if p.suffix == ".jsonl":
        return str(p), parse_jsonl_transcript(p)
    return str(p), p.read_text(encoding="utf-8", errors="ignore")


def embed(text: str) -> list[float]:
    cmd = [str(EMBED_BIN), "embed", "--text", text] if EMBED_BIN.exists() else ["python3", str(EMBED_SCRIPT), "embed", "--text", text]
    proc = sh(cmd, timeout=120)
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or proc.stdout.strip() or "embedding failed")
    payload = json.loads(proc.stdout or "{}")
    vectors = payload.get("vectors") or []
    if not vectors:
        raise RuntimeError("no embedding returned")
    return [float(x) for x in vectors[0]]


def vec_sql(vec: list[float]) -> str:
    return "'[%s]'::vector" % ",".join(f"{x:.8f}" for x in vec)


def call_ollama(prompt: str, model: str) -> dict[str, Any]:
    payload = {
        "model": model,
        "prompt": prompt,
        "stream": False,
        "format": "json",
        "options": {"temperature": 0.1},
    }
    proc = sh(["curl", "-sS", "http://127.0.0.1:11434/api/generate", "-d", json.dumps(payload)], timeout=180)
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or "ollama call failed")
    wrapped = json.loads(proc.stdout or "{}")
    response = (wrapped.get("response") or "{}").strip()
    try:
        return json.loads(response)
    except json.JSONDecodeError:
        start = response.find("{")
        end = response.rfind("}")
        if start >= 0 and end > start:
            try:
                return json.loads(response[start : end + 1])
            except json.JSONDecodeError:
                return {"facts": []}
        return {"facts": []}


def normalize_fact(raw: dict[str, Any]) -> AtomicFact | None:
    try:
        fact_type = str(raw.get("type", "")).strip()
        content = " ".join(str(raw.get("content", "")).split()).strip()
        tags = [str(t).strip() for t in raw.get("tags", []) if str(t).strip()]
        people = [str(p).strip() for p in raw.get("people", []) if str(p).strip()]
        projects = [str(p).strip() for p in raw.get("projects", []) if str(p).strip()]
        importance = max(0.0, min(1.0, float(raw.get("importance", 0.5))))
        confidence = max(0.0, min(1.0, float(raw.get("confidence", 0.5))))
    except Exception:
        return None

    if fact_type not in VALID_TYPES:
        return None
    if len(content) < 8:
        return None

    supersedes_id = raw.get("supersedes_id")
    if supersedes_id is not None:
        try:
            supersedes_id = int(supersedes_id)
        except Exception:
            supersedes_id = None

    return AtomicFact(
        fact_type=fact_type,
        content=content,
        tags=tags[:12],
        people=people[:12],
        projects=projects[:12],
        importance=importance,
        confidence=confidence,
        supersedes_id=supersedes_id,
    )


def find_neighbors(vec: list[float]) -> list[tuple[int, str, float]]:
    sql = f"""
SELECT id, object_value, 1 - (embedding_local <=> {vec_sql(vec)}) AS similarity
FROM cortana_memory_semantic
WHERE active = TRUE
  AND embedding_local IS NOT NULL
  AND (1 - (embedding_local <=> {vec_sql(vec)})) >= 0.85
ORDER BY embedding_local <=> {vec_sql(vec)}
LIMIT 5;
"""
    rows = psql(sql, capture=True).splitlines()
    out: list[tuple[int, str, float]] = []
    for row in rows:
        if not row:
            continue
        rid, text, sim = row.split("|", 2)
        out.append((int(rid), text, float(sim)))
    return out


def classify_meaning(existing: str, new: str, model: str) -> tuple[bool, bool]:
    prompt = (
        "Return only JSON {\"identical\":true|false,\"updated\":true|false}. "
        "identical=true when both statements mean the same thing. "
        "updated=true when B should supersede A (same subject, newer/corrected detail).\n"
        f"A: {existing}\nB: {new}\n"
    )
    try:
        parsed = call_ollama(prompt, model=model)
    except Exception:
        return False, False
    return bool(parsed.get("identical")), bool(parsed.get("updated"))


def insert_fact(fact: AtomicFact, source_ref: str, vec: list[float], supersedes_id: int | None, dry_run: bool) -> dict[str, Any]:
    metadata = json.dumps({
        "atomic_fact": True,
        "importance": fact.importance,
        "tags": fact.tags,
        "people": fact.people,
        "projects": fact.projects,
    })

    if dry_run:
        return {"action": "would_insert", "fact": fact.content, "supersedes_id": supersedes_id}

    sql = f"""
INSERT INTO cortana_memory_semantic (
  fact_type, subject, predicate, object_value,
  confidence, trust, stability,
  first_seen_at, last_seen_at,
  source_type, source_ref, fingerprint,
  metadata, embedding_local, embedding_model,
  extraction_source, supersedes_id
) VALUES (
  {q(fact.fact_type)},
  'hamel',
  'stated',
  {q(fact.content)},
  {fact.confidence:.3f},
  {max(0.5, fact.confidence):.3f},
  {max(0.4, fact.importance):.3f},
  NOW(), NOW(),
  'atomic_extraction',
  {q(source_ref)},
  md5({q(fact.fact_type + '|' + fact.content)}),
  {q(metadata)}::jsonb,
  {vec_sql(vec)},
  'BAAI/bge-small-en-v1.5',
  {q(source_ref)},
  {str(supersedes_id) if supersedes_id else 'NULL'}
)
ON CONFLICT (fact_type, subject, predicate, object_value)
DO UPDATE SET
  last_seen_at = NOW(),
  confidence = GREATEST(cortana_memory_semantic.confidence, EXCLUDED.confidence),
  metadata = cortana_memory_semantic.metadata || EXCLUDED.metadata
RETURNING id;
"""
    new_id = int(psql(sql, capture=True))

    if supersedes_id:
        psql(
            f"UPDATE cortana_memory_semantic "
            f"SET active=FALSE, superseded_by={new_id}, superseded_at=NOW() "
            f"WHERE id={supersedes_id};"
        )

    return {"action": "insert", "id": new_id, "fact": fact.content, "supersedes_id": supersedes_id}


def process_text(text: str, source_ref: str, model: str, dry_run: bool) -> list[dict[str, Any]]:
    prompt_template = load_prompt()
    prompt = prompt_template.replace("{{TRANSCRIPT}}", text[:18000]).replace("{{EXISTING_FACTS}}", "[]")
    parsed = call_ollama(prompt, model=model)
    facts_raw = parsed.get("facts", []) if isinstance(parsed, dict) else []

    results: list[dict[str, Any]] = []
    for item in facts_raw:
        fact = normalize_fact(item)
        if not fact:
            continue

        vec = embed(fact.content)
        neighbors = find_neighbors(vec)
        best = neighbors[0] if neighbors else None

        if best and best[2] > 0.95:
            identical, updated = classify_meaning(best[1], fact.content, model)
            if identical:
                results.append({"action": "skip_duplicate", "existing_id": best[0], "similarity": round(best[2], 4), "fact": fact.content})
                continue
            if updated:
                results.append(insert_fact(fact, source_ref, vec, best[0], dry_run=dry_run))
                continue

        if best and 0.85 <= best[2] <= 0.95:
            identical, updated = classify_meaning(best[1], fact.content, model)
            if identical:
                results.append({"action": "skip_duplicate", "existing_id": best[0], "similarity": round(best[2], 4), "fact": fact.content})
                continue
            if updated:
                results.append(insert_fact(fact, source_ref, vec, best[0], dry_run=dry_run))
                continue

        # Similarity threshold for semantic match check requested by spec.
        if best and best[2] >= 0.92:
            results.append(insert_fact(fact, source_ref, vec, None, dry_run=dry_run))
        else:
            results.append(insert_fact(fact, source_ref, vec, None, dry_run=dry_run))

    return results


def cmd_extract(args: argparse.Namespace) -> None:
    ensure_schema()
    source_ref, text = read_input(args.input)
    results = process_text(text, source_ref=source_ref, model=args.model, dry_run=args.dry_run)
    print(json.dumps({"ok": True, "mode": "extract", "source": source_ref, "count": len(results), "results": results}, indent=2))


def recent_session_files(since_hours: int) -> list[Path]:
    import glob

    cutoff = datetime.now(timezone.utc) - timedelta(hours=since_hours)
    out: list[Path] = []
    for raw in glob.glob(SESSIONS_GLOB):
        p = Path(raw)
        if ".deleted." in p.name:
            continue
        try:
            mtime = datetime.fromtimestamp(p.stat().st_mtime, tz=timezone.utc)
        except FileNotFoundError:
            continue
        if mtime >= cutoff:
            out.append(p)
    return sorted(out)


def cmd_batch(args: argparse.Namespace) -> None:
    ensure_schema()
    files = recent_session_files(args.since_hours)
    all_results: list[dict[str, Any]] = []
    for f in files:
        text = parse_jsonl_transcript(f)
        if not text.strip():
            continue
        all_results.extend(process_text(text, source_ref=str(f), model=args.model, dry_run=args.dry_run))

    print(json.dumps({
        "ok": True,
        "mode": "batch",
        "since_hours": args.since_hours,
        "sessions": len(files),
        "result_count": len(all_results),
        "results": all_results,
    }, indent=2))


def parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Extract atomic facts into cortana_memory_semantic")
    sub = p.add_subparsers(dest="cmd", required=True)

    p_extract = sub.add_parser("extract", help="Extract facts from one file or stdin")
    p_extract.add_argument("--input", required=True, help="Input file path or '-' for stdin")
    p_extract.add_argument("--dry-run", action="store_true")
    p_extract.add_argument("--model", default="phi3:mini")
    p_extract.set_defaults(func=cmd_extract)

    p_batch = sub.add_parser("batch", help="Process recent session logs")
    p_batch.add_argument("--since-hours", type=int, default=24)
    p_batch.add_argument("--dry-run", action="store_true")
    p_batch.add_argument("--model", default="phi3:mini")
    p_batch.set_defaults(func=cmd_batch)

    return p


def main() -> None:
    args = parser().parse_args()
    args.func(args)


if __name__ == "__main__":
    main()

File: tools/memory/extract_facts.ts
#!/usr/bin/env npx tsx

import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";
import { resolveRepoPath } from "../lib/paths.js";

const WORKSPACE = resolveRepoPath();
const PSQL_BIN = "/opt/homebrew/opt/postgresql@17/bin/psql";
const DB_NAME = "cortana";
const PROMPT_FILE = path.join(WORKSPACE, "tools", "memory", "prompts", "fact_extraction.txt");
const EMBED_SCRIPT = path.join(WORKSPACE, "tools", "embeddings", "embed.py");
const EMBED_BIN = path.join(WORKSPACE, "tools", "embeddings", "embed");
const SESSIONS_GLOB = path.join("~", ".openclaw", "agents", "main", "sessions", "*.jsonl");

const VALID_TYPES = new Set(["fact", "preference", "event", "system_rule"]);

type AtomicFact = {
  fact_type: string;
  content: string;
  tags: string[];
  people: string[];
  projects: string[];
  importance: number;
  confidence: number;
  supersedes_id: number | null;
};

function sh(cmd: string[], timeout = 120000): ReturnType<typeof spawnSync> {
  return spawnSync(cmd[0], cmd.slice(1), { encoding: "utf8", timeout });
}

function psql(sql: string, capture = false): string {
  const proc = sh([PSQL_BIN, DB_NAME, "-q", "-v", "ON_ERROR_STOP=1", "-At", "-c", sql], 180000);
  if (proc.status !== 0) {
    throw new Error((proc.stderr || "").trim() || "psql failed");
  }
  return capture ? (proc.stdout || "").trim() : "";
}

function q(value: unknown): string {
  if (value === null || value === undefined) return "NULL";
  return `'${String(value).replace(/'/g, "''")}'`;
}

function ensureSchema(): void {
  const sql = `
ALTER TABLE cortana_memory_semantic
  ADD COLUMN IF NOT EXISTS fact_type TEXT,
  ADD COLUMN IF NOT EXISTS superseded_by BIGINT,
  ADD COLUMN IF NOT EXISTS superseded_at TIMESTAMPTZ,
  ADD COLUMN IF NOT EXISTS supersedes_id BIGINT,
  ADD COLUMN IF NOT EXISTS extraction_source TEXT,
  ADD COLUMN IF NOT EXISTS embedding_local VECTOR(384);

DO $$
BEGIN
  IF NOT EXISTS (
    SELECT 1 FROM pg_constraint
    WHERE conname='cortana_memory_semantic_superseded_by_fkey'
  ) THEN
    ALTER TABLE cortana_memory_semantic
      ADD CONSTRAINT cortana_memory_semantic_superseded_by_fkey
      FOREIGN KEY (superseded_by) REFERENCES cortana_memory_semantic(id);
  END IF;
END $$;

DO $$
BEGIN
  IF NOT EXISTS (
    SELECT 1 FROM pg_constraint
    WHERE conname='cortana_memory_semantic_supersedes_id_fkey'
  ) THEN
    ALTER TABLE cortana_memory_semantic
      ADD CONSTRAINT cortana_memory_semantic_supersedes_id_fkey
      FOREIGN KEY (supersedes_id) REFERENCES cortana_memory_semantic(id);
  END IF;
END $$;

DO $$
BEGIN
  IF EXISTS (
    SELECT 1 FROM pg_constraint
    WHERE conname='cortana_memory_semantic_fact_type_check'
  ) THEN
    ALTER TABLE cortana_memory_semantic DROP CONSTRAINT cortana_memory_semantic_fact_type_check;
  END IF;

  ALTER TABLE cortana_memory_semantic
    ALTER COLUMN fact_type SET DEFAULT 'fact';

  UPDATE cortana_memory_semantic
  SET fact_type = 'fact'
  WHERE fact_type IS NULL;

  ALTER TABLE cortana_memory_semantic
    ADD CONSTRAINT cortana_memory_semantic_fact_type_check
    CHECK (fact_type = ANY (ARRAY['fact','preference','event','system_rule','decision','rule','relationship']));
END $$;

CREATE INDEX IF NOT EXISTS idx_memory_semantic_embedding_local_hnsw
  ON cortana_memory_semantic USING hnsw (embedding_local vector_cosine_ops)
  WHERE embedding_local IS NOT NULL;
`;
  psql(sql);
}

function loadPrompt(): string {
  if (!fs.existsSync(PROMPT_FILE)) {
    throw new Error(`Missing prompt template: ${PROMPT_FILE}`);
  }
  return fs.readFileSync(PROMPT_FILE, "utf8");
}

function parseJsonlTranscript(filePath: string): string {
  const lines: string[] = [];
  const raw = fs.readFileSync(filePath, "utf8");
  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    let obj: any;
    try {
      obj = JSON.parse(trimmed);
    } catch {
      continue;
    }
    if (obj.type !== "message") continue;
    const msg = obj.message ?? {};
    const role = msg.role;
    if (role !== "user" && role !== "assistant") continue;
    const textParts: string[] = [];
    for (const chunk of msg.content ?? []) {
      if (chunk && typeof chunk === "object" && chunk.type === "text" && typeof chunk.text === "string") {
        textParts.push(chunk.text);
      }
    }
    if (textParts.length) {
      lines.push(`${String(role).toUpperCase()}: ${textParts.join(" ").trim()}`);
    }
  }
  return lines.join("\n");
}

function readInput(inputPath: string): [string, string] {
  if (inputPath === "-") {
    return ["stdin", fs.readFileSync(0, "utf8")];
  }
  const p = path.resolve(inputPath);
  if (!fs.existsSync(p)) {
    throw new Error(inputPath);
  }
  if (p.endsWith(".jsonl")) {
    return [p, parseJsonlTranscript(p)];
  }
  return [p, fs.readFileSync(p, { encoding: "utf8", flag: "r" })];
}

function embed(text: string): number[] {
  const cmd = fs.existsSync(EMBED_BIN)
    ? [EMBED_BIN, "embed", "--text", text]
    : ["python3", EMBED_SCRIPT, "embed", "--text", text];
  const proc = sh(cmd, 120000);
  if (proc.status !== 0) {
    throw new Error((proc.stderr || "").trim() || (proc.stdout || "").trim() || "embedding failed");
  }
  const payload = JSON.parse(proc.stdout || "{}");
  const vectors = payload.vectors || [];
  if (!vectors.length) throw new Error("no embedding returned");
  return (vectors[0] as any[]).map((x) => Number(x));
}

function vecSql(vec: number[]): string {
  return `'[${vec.map((x) => x.toFixed(8)).join(",")}]'::vector`;
}

async function callOllama(prompt: string, model: string): Promise<Record<string, any>> {
  const payload = {
    model,
    prompt,
    stream: false,
    format: "json",
    options: { temperature: 0.1 },
  };

  const res = await fetch("http://127.0.0.1:11434/api/generate", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error((await res.text()) || "ollama call failed");
  }
  const wrapped = (await res.json()) as any;
  const response = String(wrapped.response ?? "{}").trim();
  try {
    return JSON.parse(response);
  } catch {
    const start = response.indexOf("{");
    const end = response.lastIndexOf("}");
    if (start >= 0 && end > start) {
      try {
        return JSON.parse(response.slice(start, end + 1));
      } catch {
        return { facts: [] };
      }
    }
    return { facts: [] };
  }
}

function normalizeFact(raw: Record<string, any>): AtomicFact | null {
  try {
    const factType = String(raw.type ?? "").trim();
    const content = String(raw.content ?? "").split(/\s+/).join(" ").trim();
    const tags = Array.isArray(raw.tags) ? raw.tags.map((t: any) => String(t).trim()).filter(Boolean) : [];
    const people = Array.isArray(raw.people) ? raw.people.map((p: any) => String(p).trim()).filter(Boolean) : [];
    const projects = Array.isArray(raw.projects) ? raw.projects.map((p: any) => String(p).trim()).filter(Boolean) : [];
    const importance = Math.max(0.0, Math.min(1.0, Number(raw.importance ?? 0.5)));
    const confidence = Math.max(0.0, Math.min(1.0, Number(raw.confidence ?? 0.5)));

    if (!VALID_TYPES.has(factType)) return null;
    if (content.length < 8) return null;

    let supersedesId: number | null = null;
    if (raw.supersedes_id !== null && raw.supersedes_id !== undefined) {
      const v = Number.parseInt(String(raw.supersedes_id), 10);
      supersedesId = Number.isFinite(v) ? v : null;
    }

    return {
      fact_type: factType,
      content,
      tags: tags.slice(0, 12),
      people: people.slice(0, 12),
      projects: projects.slice(0, 12),
      importance,
      confidence,
      supersedes_id: supersedesId,
    };
  } catch {
    return null;
  }
}

function findNeighbors(vec: number[]): Array<[number, string, number]> {
  const sql = `
SELECT id, object_value, 1 - (embedding_local <=> ${vecSql(vec)}) AS similarity
FROM cortana_memory_semantic
WHERE active = TRUE
  AND embedding_local IS NOT NULL
  AND (1 - (embedding_local <=> ${vecSql(vec)})) >= 0.85
ORDER BY embedding_local <=> ${vecSql(vec)}
LIMIT 5;
`;
  const rows = psql(sql, true).split(/\r?\n/);
  const out: Array<[number, string, number]> = [];
  for (const row of rows) {
    if (!row) continue;
    const [rid, text, sim] = row.split("|", 3);
    out.push([Number(rid), text, Number(sim)]);
  }
  return out;
}

async function classifyMeaning(existing: string, neu: string, model: string): Promise<[boolean, boolean]> {
  const prompt =
    "Return only JSON {\"identical\":true|false,\"updated\":true|false}. " +
    "identical=true when both statements mean the same thing. " +
    "updated=true when B should supersede A (same subject, newer/corrected detail).\n" +
    `A: ${existing}\nB: ${neu}\n`;
  try {
    const parsed = await callOllama(prompt, model);
    return [Boolean(parsed.identical), Boolean(parsed.updated)];
  } catch {
    return [false, false];
  }
}

function insertFact(fact: AtomicFact, sourceRef: string, vec: number[], supersedesId: number | null, dryRun: boolean): Record<string, any> {
  const metadata = JSON.stringify({
    atomic_fact: true,
    importance: fact.importance,
    tags: fact.tags,
    people: fact.people,
    projects: fact.projects,
  });

  if (dryRun) {
    return { action: "would_insert", fact: fact.content, supersedes_id: supersedesId };
  }

  const sql = `
INSERT INTO cortana_memory_semantic (
  fact_type, subject, predicate, object_value,
  confidence, trust, stability,
  first_seen_at, last_seen_at,
  source_type, source_ref, fingerprint,
  metadata, embedding_local, embedding_model,
  extraction_source, supersedes_id
) VALUES (
  ${q(fact.fact_type)},
  'hamel',
  'stated',
  ${q(fact.content)},
  ${fact.confidence.toFixed(3)},
  ${Math.max(0.5, fact.confidence).toFixed(3)},
  ${Math.max(0.4, fact.importance).toFixed(3)},
  NOW(), NOW(),
  'atomic_extraction',
  ${q(sourceRef)},
  md5(${q(`${fact.fact_type}|${fact.content}`)}),
  ${q(metadata)}::jsonb,
  ${vecSql(vec)},
  'BAAI/bge-small-en-v1.5',
  ${q(sourceRef)},
  ${supersedesId ? String(supersedesId) : "NULL"}
)
ON CONFLICT (fact_type, subject, predicate, object_value)
DO UPDATE SET
  last_seen_at = NOW(),
  confidence = GREATEST(cortana_memory_semantic.confidence, EXCLUDED.confidence),
  metadata = cortana_memory_semantic.metadata || EXCLUDED.metadata
RETURNING id;
`;
  const newId = Number(psql(sql, true));

  if (supersedesId) {
    psql(
      `UPDATE cortana_memory_semantic SET active=FALSE, superseded_by=${newId}, superseded_at=NOW() WHERE id=${supersedesId};`
    );
  }

  return { action: "insert", id: newId, fact: fact.content, supersedes_id: supersedesId };
}

async function processText(text: string, sourceRef: string, model: string, dryRun: boolean): Promise<Record<string, any>[]> {
  const promptTemplate = loadPrompt();
  const prompt = promptTemplate.replace("{{TRANSCRIPT}}", text.slice(0, 18000)).replace("{{EXISTING_FACTS}}", "[]");
  const parsed = await callOllama(prompt, model);
  const factsRaw = typeof parsed === "object" && parsed ? parsed.facts ?? [] : [];

  const results: Record<string, any>[] = [];
  for (const item of factsRaw) {
    const fact = normalizeFact(item);
    if (!fact) continue;

    const vec = embed(fact.content);
    const neighbors = findNeighbors(vec);
    const best = neighbors.length ? neighbors[0] : null;

    if (best && best[2] > 0.95) {
      const [identical, updated] = await classifyMeaning(best[1], fact.content, model);
      if (identical) {
        results.push({ action: "skip_duplicate", existing_id: best[0], similarity: Number(best[2].toFixed(4)), fact: fact.content });
        continue;
      }
      if (updated) {
        results.push(insertFact(fact, sourceRef, vec, best[0], dryRun));
        continue;
      }
    }

    if (best && best[2] >= 0.85 && best[2] <= 0.95) {
      const [identical, updated] = await classifyMeaning(best[1], fact.content, model);
      if (identical) {
        results.push({ action: "skip_duplicate", existing_id: best[0], similarity: Number(best[2].toFixed(4)), fact: fact.content });
        continue;
      }
      if (updated) {
        results.push(insertFact(fact, sourceRef, vec, best[0], dryRun));
        continue;
      }
    }

    results.push(insertFact(fact, sourceRef, vec, null, dryRun));
  }

  return results;
}

async function cmdExtract(input: string, model: string, dryRun: boolean): Promise<void> {
  ensureSchema();
  const [sourceRef, text] = readInput(input);
  const results = await processText(text, sourceRef, model, dryRun);
  console.log(
    JSON.stringify({ ok: true, mode: "extract", source: sourceRef, count: results.length, results }, null, 2)
  );
}

function expandTilde(input: string): string {
  if (input.startsWith("~/")) {
    return path.join(process.env.HOME || "", input.slice(2));
  }
  return input;
}

function recentSessionFiles(sinceHours: number): string[] {
  const globPath = expandTilde(SESSIONS_GLOB);
  const dir = path.dirname(globPath);
  const base = path.basename(globPath).replace(/\./g, "\\.").replace(/\*/g, ".*");
  const matcher = new RegExp(`^${base}$`);
  const cutoff = Date.now() - sinceHours * 60 * 60 * 1000;
  const out: string[] = [];
  let entries: string[] = [];
  try {
    entries = fs.readdirSync(dir);
  } catch {
    return [];
  }
  for (const name of entries) {
    if (name.includes(".deleted.")) continue;
    if (!matcher.test(name)) continue;
    const full = path.join(dir, name);
    try {
      const mtime = fs.statSync(full).mtimeMs;
      if (mtime >= cutoff) out.push(full);
    } catch {
      continue;
    }
  }
  return out.sort();
}

async function cmdBatch(sinceHours: number, model: string, dryRun: boolean): Promise<void> {
  ensureSchema();
  const files = recentSessionFiles(sinceHours);
  const allResults: Record<string, any>[] = [];
  for (const f of files) {
    const text = parseJsonlTranscript(f);
    if (!text.trim()) continue;
    allResults.push(...(await processText(text, f, model, dryRun)));
  }

  console.log(
    JSON.stringify(
      {
        ok: true,
        mode: "batch",
        since_hours: sinceHours,
        sessions: files.length,
        result_count: allResults.length,
        results: allResults,
      },
      null,
      2
    )
  );
}

function printHelp(): void {
  const text = `usage: extract_facts.ts [-h] {extract,batch} ...\n\nExtract atomic facts into cortana_memory_semantic\n\noptions:\n  -h, --help  show this help message and exit`;
  console.log(text);
}

async function main(): Promise<void> {
  const argv = process.argv.slice(2);
  if (argv.includes("-h") || argv.includes("--help") || argv.length === 0) {
    printHelp();
    if (argv.length === 0) process.exit(2);
    return;
  }

  const cmd = argv[0];
  if (cmd === "extract") {
    let input: string | null = null;
    let dryRun = false;
    let model = "phi3:mini";
    for (let i = 1; i < argv.length; i += 1) {
      const arg = argv[i];
      const next = argv[i + 1];
      if (arg === "--input" && next) {
        input = next;
        i += 1;
      } else if (arg === "--dry-run") {
        dryRun = true;
      } else if (arg === "--model" && next) {
        model = next;
        i += 1;
      } else if (arg === "-h" || arg === "--help") {
        printHelp();
        return;
      } else if (arg.startsWith("-")) {
        console.error(`Unknown argument: ${arg}`);
        printHelp();
        process.exit(2);
      }
    }
    if (!input) {
      console.error("--input is required");
      process.exit(2);
    }
    await cmdExtract(input, model, dryRun);
    return;
  }

  if (cmd === "batch") {
    let sinceHours = 24;
    let dryRun = false;
    let model = "phi3:mini";
    for (let i = 1; i < argv.length; i += 1) {
      const arg = argv[i];
      const next = argv[i + 1];
      if (arg === "--since-hours" && next) {
        sinceHours = Number.parseInt(next, 10);
        i += 1;
      } else if (arg === "--dry-run") {
        dryRun = true;
      } else if (arg === "--model" && next) {
        model = next;
        i += 1;
      } else if (arg === "-h" || arg === "--help") {
        printHelp();
        return;
      } else if (arg.startsWith("-")) {
        console.error(`Unknown argument: ${arg}`);
        printHelp();
        process.exit(2);
      }
    }
    await cmdBatch(sinceHours, model, dryRun);
    return;
  }

  console.error("Unknown command");
  process.exit(2);
}

main().catch((err) => {
  const msg = err instanceof Error ? err.message : String(err);
  console.error(msg);
  process.exit(1);
});
